libs/agent-sm/agent-policy/src/main/java/org/opensearch/secure_sm/policy/PolicyFile.java

```diff
@@ -0,0 +1,311 @@
+            synchronized (pc) {
+                Enumeration<Permission> e = pc.elements();
+                while (e.hasMoreElements()) {
+                    perms.add(e.nextElement());
+                }
+            }
+        }
+
+        return perms;
+    }
+
+    @Override
+    public PermissionCollection getPermissions(CodeSource codesource) {
+        if (codesource == null) {
+            return new Permissions();
+        }
+
+        Permissions perms = new Permissions();
+        CodeSource canonicalCodeSource = canonicalizeCodebase(codesource);
+
+        for (PolicyEntry entry : policyInfo.policyEntries) {
+            if (entry.getCodeSource().implies(canonicalCodeSource)) {
+                for (Permission permission : entry.permissions) {
+                    perms.add(permission);
+                }
+            }
+        }
+
+        return perms;
+    }
+
+    public static InputStream getInputStream(URL url) throws IOException {
+        if ("file".equals(url.getProtocol())) {
+            String path = url.getFile().replace('/', File.separatorChar);
+            path = URLDecoder.decode(path, StandardCharsets.UTF_8);
+            return new FileInputStream(path);
+        } else {
+            return url.openStream();
+        }
+    }
+
+    /**
+     * Given a GrantEntry, create a codeSource.
+     *
+     * @return null if signedBy alias is not recognized
+     */
+    private CodeSource getCodeSource(GrantNode grantEntry, PolicyInfo newInfo) throws java.net.MalformedURLException {
+        Certificate[] certs = null;
+        URL location = (grantEntry.codeBase != null) ? newURL(grantEntry.codeBase) : null;
+
+        return (canonicalizeCodebase(new CodeSource(location, certs)));
+    }
+
+    private void addGrantNode(GrantNode grantEntry, PolicyInfo newInfo) throws Exception {
+
+        CodeSource codesource = getCodeSource(grantEntry, newInfo);
+        if (codesource == null) return;
+
+        PolicyEntry entry = new PolicyEntry(codesource);
+        Enumeration<PermissionNode> enum_ = grantEntry.permissionElements();
+        while (enum_.hasMoreElements()) {
+            PermissionNode pe = enum_.nextElement();
+            try {
+                expandPermissionName(pe);
+
+                Optional<Permission> perm = getInstance(pe.permission, pe.name, pe.action);
+
+                // add the permission to policy entry.
+                perm.ifPresent(entry::add);
+            } catch (ClassNotFoundException cfne) {
+                cfne.printStackTrace(System.err);
+            }
+        }
+        newInfo.policyEntries.add(entry);
+    }
+
+    private static Optional<Permission> getInstance(String type, String name, String actions) throws ClassNotFoundException {
+        Class<?> pc = Class.forName(type, false, null);
+        return Optional.ofNullable(WellKnownPermissions.getWellKnownPermission(pc, name, actions));
+    }
+
+    private void expandPermissionName(PermissionNode pe) {
```

Comment:
I have not seen this feature used in a policy file in the default distribution. Any idea what the purpose of this method is?

Reply:
I will have to take a look. but there was a couple of test failing in absence of this.

Reply:
This is resolved. See https://github.com/opensearch-project/OpenSearch/pull/17753#discussion_r2029293010